File: Python3/ffbsolver.py
#!/usr/bin/env python3
#-*- coding: utf-8 -*-

# Fibonacci "Fizz Buzz" Solver.
# Generate the first n Fibonacci numbers F(n), printing
#     "Buzz" when F(n) is divisible by 3.
#     "Fizz" when F(n) is divisible by 5.
#     "FizzBuzz" when F(n) is divisible by 15.
#     "BuzzFizz" when F(n) is prime.
#   the value F(n) otherwise.

import argparse
import sys
from math import sqrt
import pdb

def oddSeries():
    "Generate all odd numbers, starting at 1."
    odd = 1
    while True:
        yield odd
        odd += 2

class Primes:
    "Progressively grown list of prime numbers."
    def __init__(self):
        self.primes = [ 2, 3 ]
        self.lastodd = self.primes[-1]
        self.odds = oddSeries()
        got = next(self.odds)
        while self.lastodd > got:
            got = next(self.odds)

    def isPrime(self, candidate):
        "Return True if candidate is prime, else False."
        if candidate == 1:
            return False
        if candidate <= self.primes[-1] and candidate in self.primes:
            return True
        for divisor in self.primes:
            if candidate % divisor == 0:
                # not prime
                return False
        # got here? then it's prime.
        return True

    def stretch(self, limit):
        "Stretch list of primes to given limit."

        if limit < self.lastodd:
            return  # too soon to stretch
        while True:
            self.lastodd = next(self.odds)
            # print ("self.lastodd", self.lastodd)  # very verbose
            if self.isPrime(self.lastodd):
                self.primes.append(self.lastodd)
            if self.lastodd >= limit:
                return

def classify(value, primes, verbose=False):
    """Generate messages for matched criteria:
    * "Buzz" when F(n) is divisible by 3.
    * "Fizz" when F(n) is divisible by 5.
    * "FizzBuzz" when F(n) is divisible by 15.
    * "BuzzFizz" when F(n) is prime.
    * the value F(n), if none of the above.
    """

    primes.stretch(int(sqrt(value)))

    vv = verbose
    msgs = []
    # look for desired divisors
    if value % 5 == 0:  # divisible by 5
        msgs.append("Fizz" + ["","5"][vv])
        #msgs.append("Fizz" + verbose?"":"5")
    if value % 3 == 0:  # divisible by 3
        msgs.append("Buzz" + ["","3"][vv])
    if value % 15 == 0: # divisible by 15
        msgs.append("FizzBuzz" + ["","15"][vv])
    if primes.isPrime(value):
        msgs.append ("BuzzFizz" + ["","P"][vv])

    if msgs:
        msg = ' '.join(msgs)
    else:
        msg = "%d" % value

    return msg


def fibsequence(N):
    """Generate sequence of Fibonacci numbers up to the
    'N'th member of the sequence."""

    x = 0
    y = 1
    count = 0
    while count < N:
        x, y = y, x + y
        count += 1
        yield x
        # x = F(count)


def parseargs(argv):
    "Parse command line arguments."

    desc = "Fibonacci Fizz Buzz solver"
    parser = argparse.ArgumentParser(description=desc)
    parser.add_argument('N', type=int,
        help="Last Fibonacci number to consider")
    parser.add_argument('-v', '--verbose', help="enable verbose output",
        action="store_true")

    args = parser.parse_args()
    return args


def main(argv):

    args = parseargs(argv)

    primes = Primes()

    for fn in fibsequence(args.N):
        # fn is F(n)
        if args.verbose:
            print(fn, end=' ')          # prepend result by number and blank.
        print(classify(fn, primes, verbose=args.verbose))


sys.exit(main(sys.argv))

# vim: set sw=4 tw=80 :

File: C++/ffbsolver.cc
//
//  main.cpp
//  swiftfib-c++
//
//  Created by Rick Kwan on 9/13/17.
//  Copyright © 2017 Rick Kwan. All rights reserved.
//

#include <iostream>
#include <sstream>
#include <cmath>
#include <vector>
#include <unistd.h>
#include <getopt.h>

using namespace std;

struct GlblArgs_t {
    bool verbose;               // be verbose, or not.
    int nFibs;                  // number of fibbonacci numbers
} glblargs;

static const char* optstr = "vh?"; // options string

class Primes {
  private:
    vector <long> primeslist;
  public:
    Primes(int sz);
    ~Primes();
    bool isPrime(long candidate);
    void stretch(long limit);
  private:
    long lastodd;     // some recent odd number
};

Primes::Primes(int sz)
{
    primeslist.reserve(sz);
    primeslist.push_back(2);
    primeslist.push_back(3);
    lastodd = *primeslist.end();
}

Primes::~Primes() { }

bool
Primes::isPrime(long candidate)
{
    if (candidate == 1)
        return false;

    std::vector<long>::iterator it;
    long lastprime = primeslist.back();
    if (candidate <= lastprime) {
        for (it=primeslist.begin(); it != primeslist.end(); ++it)
            if (candidate == *it)
                return true;
    }
    long divisor;
    std::vector<long>::iterator itdiv;
    for (itdiv=primeslist.begin(); itdiv != primeslist.end(); ++itdiv) {
        divisor = *itdiv;
        if (candidate % divisor == 0)
            return false;
    }
    // got here? then candidate is prime
    return true;
}

void
Primes::stretch(long limit)
{
    if (limit < lastodd)
        return;
    while (true) {
        lastodd += 2;
        if (isPrime(lastodd))
            primeslist.push_back(lastodd);
        if (lastodd >= limit)
            return;
    }
}


long
figseq_next()
{
    static long x = 0;      // current Fib value to return
    static long y = 1;      // next Fib value
    long newy = x + y;
    x = y;
    y = newy;
    return x;
}

string
classify(long value, Primes* primes, bool verbose=false)
{
    vector <string> msgs;
    std::vector<string>::iterator itmsgs;

    primes->stretch(int(sqrt(value)));
    bool vv = verbose;

    if (value  % 5 == 0)            // divisible by 5
        msgs.push_back("Fizz" + string(vv?"5":""));
    if (value % 3 == 0)             // divisible by 3
        msgs.push_back("Buzz" + string(vv?"3":""));
    if (value % 15 == 0)            // divisible by 15
        msgs.push_back("FizzBuzz" + string(vv?"15":""));
    if (primes->isPrime(value))      // check for primeness
        msgs.push_back("BuzzFizz" + string(vv?"P":""));

    string msgline = "";
    if (msgs.size() != 0) {
        string first = *msgs.begin();
        msgline = string(first);
        msgs.erase(msgs.begin());
        for (itmsgs = msgs.begin(); itmsgs != msgs.end(); ++itmsgs) {
            msgline += string(" ") + string(*itmsgs);
        }
    } else {
        msgline = std::to_string(value);
    }
    return msgline;
}

void
displayUsage(void)
{
    printf( "ffbsolver - Fibonacci fizz buzz solver\n" );
    exit (EXIT_FAILURE);
}



int
main(int argc, char** argv) {

    int opt;
    extern char *optarg;
    extern int optind, optopt;

    glblargs.verbose = false;
    glblargs.nFibs = 0;

    while ((opt = getopt(argc, argv, optstr)) != -1) {
        switch (opt) {
          case 'v':
            glblargs.verbose = true;
            break;
          case 'h':
          case '?':
            displayUsage();
            break;
          default:
            /* can't get here */
            break;
        }
        opt = getopt( argc, argv, optstr);
    }
    for ( ; optind < argc; optind++) {
        glblargs.nFibs = stoi(argv[optind]);
    }

    Primes* primes = new Primes(glblargs.nFibs);

    for (int n=0; n<glblargs.nFibs; ++n) {
        long fn = figseq_next();
        if (glblargs.verbose)
            std::cout << fn << " " ;
        std::cout << classify(fn, primes, glblargs.verbose) << std::endl;
    }

    return 0;
}
